src/config_parser.rs

```diff
@@ -0,0 +1,119 @@
+use config::{Config, ConfigError, ValueKind};
+
+use crate::pipewire_listener::{PortConnections, StreamConnections};
+
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub struct TurboAudioConfig {
+    pub device_name: Option<String>,
+    pub jack: bool,
+    pub sample_rate: i64,
+    pub stream_connections: Vec<StreamConnections>,
+}
+
+impl TurboAudioConfig {
+    fn new() -> Self {
+        Self {
+            device_name: None,
+            jack: false,
+            sample_rate: 48000,
+            stream_connections: vec![],
+        }
+    }
+}
+
+pub fn parse_config(config_name: &str) -> anyhow::Result<TurboAudioConfig> {
+    let mut config = TurboAudioConfig::new();
+
+    let settings = Config::builder()
+        .add_source(config::File::with_name(config_name))
+        .build()
+        .unwrap();
+
+    read_optional_variable(&mut config.jack, settings.get_bool("jack"))?;
+    read_optional_variable(&mut config.sample_rate, settings.get_int("sample_rate"))?;
+
+    let mut device_name = "Default Input Device".to_string();
+    if read_optional_variable(&mut device_name, settings.get_string("device_name"))? {
```

Comment:
can't you pass in a `&mut` to `config.device_name` directly in `read_optional_variable`? To have a default values you could use `or_else` on `Result`. This would impact `read_optional_variable`. Imo, this function could simply be removed and we could use `or_else` for all the signing when there is a default value and simply use the `?` operator directly on the `get_x()` method of `settings`. Ex: `config.jack = settings.get_bool("jack")?;` or `config.device_name = settings.get_string("device_name").or_else("Default Input Device");`

Reply:
Non pcq le but de la fonction read_optional_variable cest de pas crash si la variable est pas dans la config, et si la variable est pas dans la config ca doit retourner un `Ok` pcq on veut pas early return pcq cest pas vrm une erreur.
Par contre on veut que ca foire si tas mis genre `"jack": 23` dans ta config pcq c'est clairement une erreur, on veut un `bool`. J'ai fait une fonction pour pas devoir repeter cette logique la.
Le code que tu proposes est pas le meme behaviour

Reply:
Not a fan of the function checking that the value is the right type if it exists and assigning if it does. Je ferais une autre fn qui check si il existe et le type check si il existe. Et une autre fn qui va read et return la valeur ex `is_valid_setting` qui check l'existance et type et raise une erreur si cest pas le bon type. Apres `config.x = setting.get_x()`. Tbh j'y tien pas plus que ca cest juste un nitpick

Reply:
Jpx essayer de faire dequoi

Reply:
Pourquoi pas juste y aller avec une fonction pure? tu retourne `Result<T, ConfigError>` ???

Reply:
nvm j'vois ce que tu fais et ca me tue en dedans un peu